src/preload.py

```diff
@@ -0,0 +1,329 @@
+from math import ceil, floor
+from requests import get
+from sh import (
+    btrfs,
+    docker,
+    dockerd,
+    e2fsck,
+    kill,
+    losetup,
+    mount,
+    parted,
+    resize2fs,
+    umount,
+    inotifywait,
+    ErrorReturnCode,
+)
+from shutil import copyfile, rmtree
+from sys import exit
+from tempfile import mkdtemp
+
+os.environ["LANG"] = "C"
+
+API_TOKEN = os.environ["API_TOKEN"]
+API_KEY = os.environ["API_KEY"]
+APP_ID = os.environ["APP_ID"]
+
+API_HOST = os.environ["API_HOST"] or "https://api.resin.io"
+REGISTRY_HOST = os.environ["REGISTRY_HOST"] or "registry2.resin.io"
+
+
+def human_size(size, precision=2):
+    suffixes = ['', 'Ki', 'Mi', 'Gi', 'Ti']
+    idx = 0
+    while idx < 4 and size >= 1024:
+        idx += 1
+        size /= 1024
+    result = "{:.{}f}".format(size, precision).rstrip("0").rstrip(".")
+    return "{} {}B".format(result, suffixes[idx])
+
+
+def log(*args):
+    print(*args, flush=True)
+
+
+def get_offset_and_size(image, partition):
+    output = parted("-s", "-m", image, "unit", "B", "p").stdout.decode("utf8")
+    lines = output.strip().split("\n")
+    for line in lines:
+        if line.startswith("{}:".format(partition)):
+            data = line.split(":")
+            offset = int(data[1][:-1])
+            size = int(data[3][:-1])
+            return offset, size
+
+
+def mount_partition(image, partition, extra_options):
+    offset, size = get_offset_and_size(image, partition)
+    mountpoint = mkdtemp()
+    options = "offset={},sizelimit={}".format(offset, size)
+    if extra_options:
+        options += "," + extra_options
+    mount("-o", options, image, mountpoint)
+    return mountpoint
+
+
+@contextmanager
+def mount_context_manager(image, partition, extra_options=""):
+    mountpoint = mount_partition(image, partition, extra_options)
+    yield mountpoint
+    umount(mountpoint)
+    os.rmdir(mountpoint)
+
+
+def losetup_partition(image, partition):
+    offset, size = get_offset_and_size(image, partition)
+    device = losetup("-f").stdout.strip()
+    losetup("-o", offset, "--sizelimit", size, device, image)
+    return device
+
+
+@contextmanager
+def losetup_context_manager(image, partition):
+    device = losetup_partition(image, partition)
+    yield device
+    losetup("-d", device)
+
+
+def api(endpoint, params=None, headers=None):
+    url = "{}/{}".format(API_HOST, endpoint)
+    params = params or {}
+    headers = headers or {}
+    if API_TOKEN:
+        headers["Authorization"] = "Bearer {}".format(API_TOKEN)
+    elif API_KEY:
+        params["apikey"] = API_KEY
+    response = get(url, params=params, headers=headers)
+    response.raise_for_status()
+    return response.json()
+
+
+def get_registry_token(image_repo):
+    params = {
+        "service": REGISTRY_HOST,
+        "scope": "repository:{}:pull".format(image_repo),
+    }
+    return api("auth/v1/token", params)["token"]
+
+
+def registry(endpoint, registry_token, headers=None, json=True):
```

Comment:
Probably rename this `json` to something else, since it redefining name 'json' from outer scope